File: roadmap-checkpoints/CassaV6-roadmap-checkpoint-MIG006-APPLIED-20260901/ROADMAP_V6/BLUETOOTH/CASSAV6_BLUETOOTH_PROTOCOL_20260820/raspberry/src/session/GattReliableDataPlaneV1.ts
import {
  RELIABLE_FRAME_TYPES,
  decodeReliableFragmentV1,
  type ReliableFrameType,
  type ReliableMessageV1
} from "../protocol/FrameCodec.js";
import {
  ReliableChannelV1,
  type ReliableChannelStoreV1
} from "../protocol/ReliableChannel.js";
import type { GattHelloExchangeV1 } from "./GattHelloExchangeV1.js";

export const GATT_RELIABLE_TRANSMITTERS = Object.freeze({
  DATA: "dataTx",
  ACK: "ackTx"
} as const);

export type GattReliableTransmitter =
  (typeof GATT_RELIABLE_TRANSMITTERS)[keyof typeof GATT_RELIABLE_TRANSMITTERS];

export class GattReliableDataPlaneError extends Error {
  readonly code: string;

  constructor(code: string, message: string, options?: ErrorOptions) {
    super(message, options);
    this.name = "GattReliableDataPlaneError";
    this.code = code;
  }
}

function fail(code: string, message: string, cause?: unknown): never {
  throw new GattReliableDataPlaneError(
    code,
    message,
    cause === undefined ? undefined : { cause }
  );
}

export class GattReliableDataPlaneV1 {
  readonly #enabled: boolean;
  readonly #helloExchange: GattHelloExchangeV1;
  readonly #store: ReliableChannelStoreV1;
  #publish: ((output: Readonly<{
    transmitter: GattReliableTransmitter;
    value: Buffer;
  }>) => void | Promise<void>) | null;
  readonly #onMessage: (
    message: ReliableMessageV1
  ) => void | Promise<void>;
  readonly #now: () => number;
  readonly #subscriptions = new Set<GattReliableTransmitter>();
  #channel: ReliableChannelV1 | null = null;
  #devicePath: string | null = null;
  #peerTrustId: string | null = null;
  #receivedFragments = 0;
  #publishedFragments = 0;
  #sessionBinds = 0;
  #resets = 0;
  #failures = 0;

  constructor(input: {
    readonly enabled: boolean;
    readonly helloExchange: GattHelloExchangeV1;
    readonly store: ReliableChannelStoreV1;
    readonly publish?: (output: Readonly<{
      transmitter: GattReliableTransmitter;
      value: Buffer;
    }>) => void | Promise<void>;
    readonly onMessage: (
      message: ReliableMessageV1
    ) => void | Promise<void>;
    readonly now?: () => number;
  }) {
    this.#enabled = input.enabled;
    this.#helloExchange = input.helloExchange;
    this.#store = input.store;
    this.#publish = input.publish ?? null;
    this.#onMessage = input.onMessage;
    this.#now = input.now ?? Date.now;
    if (this.#enabled && !this.#helloExchange.directControlEnabled) {
      fail(
        "INVALID_DATA_PLANE_CONFIGURATION",
        "reliable data requires authenticated direct control"
      );
    }
  }

  get enabled(): boolean {
    return this.#enabled;
  }

  setPublisher(
    publisher: ((output: Readonly<{
      transmitter: GattReliableTransmitter;
      value: Buffer;
    }>) => void | Promise<void>) | null
  ): void {
    if (publisher !== null && typeof publisher !== "function") {
      fail("INVALID_PUBLISHER", "data-plane publisher must be a function");
    }
    this.#publish = publisher;
  }

  setSubscription(
    transmitter: GattReliableTransmitter,
    enabled: boolean
  ): void {
    if (!Object.values(GATT_RELIABLE_TRANSMITTERS).includes(transmitter)) {
      fail("INVALID_TRANSMITTER", "GATT transmitter is not assigned");
    }
    if (!this.#enabled) {
      fail("DATA_PLANE_DISABLED", "reliable GATT data plane is disabled");
    }
    if (enabled) this.#subscriptions.add(transmitter);
    else this.#subscriptions.delete(transmitter);
  }

  async receive(devicePath: string, value: Uint8Array): Promise<Readonly<{
    complete: boolean;
    delivered: boolean;
    duplicate: boolean;
  }>> {
    this.#assertEnabled();
    if (!this.#subscriptions.has(GATT_RELIABLE_TRANSMITTERS.ACK)) {
      fail("ACK_SUBSCRIPTION_REQUIRED", "subscribe to ackTx before writing dataRx");
    }
    const channel = this.#bind(devicePath);
    this.#receivedFragments += 1;
    try {
      return await channel.receiveFragment(value);
    } catch (error) {
      this.#failures += 1;
      throw error;
    }
  }

  async send(input: {
    readonly devicePath: string;
    readonly type: ReliableFrameType;
    readonly payload: Uint8Array;
    readonly durable?: boolean;
    readonly ttlMs?: number;
  }): Promise<Readonly<{ messageId: string; durableCommitted: boolean }>> {
    this.#assertEnabled();
    if (input.type === RELIABLE_FRAME_TYPES.ACK) {
      fail("ACK_RESERVED", "ACK is emitted only by ReliableChannelV1");
    }
    if (!this.#subscriptions.has(GATT_RELIABLE_TRANSMITTERS.DATA)) {
      fail("DATA_SUBSCRIPTION_REQUIRED", "subscribe to dataTx before server send");
    }
    const channel = this.#bind(input.devicePath);
    try {
      return await channel.send({
        type: input.type,
        payload: input.payload,
        durable: input.durable,
        ttlMs: input.ttlMs
      });
    } catch (error) {
      this.#failures += 1;
      throw error;
    }
  }

  async sendBound(input: {
    readonly type: ReliableFrameType;
    readonly payload: Uint8Array;
    readonly durable?: boolean;
    readonly ttlMs?: number;
  }): Promise<Readonly<{ messageId: string; durableCommitted: boolean }>> {
    this.#assertEnabled();
    if (this.#devicePath === null || this.#channel === null) {
      return Promise.reject(
        new GattReliableDataPlaneError(
          "RELIABLE_CHANNEL_NOT_BOUND",
          "reliable data plane has no authenticated bound peer"
        )
      );
    }
    return this.send({ devicePath: this.#devicePath, ...input });
  }

  async restore(devicePath: string): Promise<number> {
    this.#assertEnabled();
    if (!this.#subscriptions.has(GATT_RELIABLE_TRANSMITTERS.DATA)) {
      fail("DATA_SUBSCRIPTION_REQUIRED", "dataTx subscription is required");
    }
    return this.#bind(devicePath).restoreDurableOutbox();
  }

  async restoreBound(): Promise<number> {
    this.#assertEnabled();
    if (this.#devicePath === null || this.#channel === null) {
      return Promise.reject(
        new GattReliableDataPlaneError(
          "RELIABLE_CHANNEL_NOT_BOUND",
          "reliable data plane has no authenticated bound peer"
        )
      );
    }
    return this.restore(this.#devicePath);
  }

  tick(): Promise<Readonly<{
    retried: number;
    suspended: number;
    expired: number;
  }>> {
    this.#assertEnabled();
    if (this.#channel === null) {
      return Promise.resolve(Object.freeze({ retried: 0, suspended: 0, expired: 0 }));
    }
    return this.#channel.tick();
  }

  reset(): void {
    this.#channel?.close();
    this.#channel = null;
    this.#devicePath = null;
    this.#peerTrustId = null;
    this.#subscriptions.clear();
    this.#resets += 1;
  }

  snapshot(): Readonly<{
    enabled: boolean;
    bound: boolean;
    dataSubscribed: boolean;
    ackSubscribed: boolean;
    receivedFragments: number;
    publishedFragments: number;
    sessionBinds: number;
    resets: number;
    failures: number;
    channel: ReturnType<ReliableChannelV1["snapshot"]> | null;
  }> {
    return Object.freeze({
      enabled: this.#enabled,
      bound: this.#channel !== null,
      dataSubscribed: this.#subscriptions.has(GATT_RELIABLE_TRANSMITTERS.DATA),
      ackSubscribed: this.#subscriptions.has(GATT_RELIABLE_TRANSMITTERS.ACK),
      receivedFragments: this.#receivedFragments,
      publishedFragments: this.#publishedFragments,
      sessionBinds: this.#sessionBinds,
      resets: this.#resets,
      failures: this.#failures,
      channel: this.#channel?.snapshot() ?? null
    });
  }

  #bind(devicePath: string): ReliableChannelV1 {
    if (this.#channel !== null) {
      if (this.#devicePath !== devicePath) {
        fail(
          "SESSION_ARBITRATION_CONFLICT",
          "one data plane cannot bind two device contexts"
        );
      }
      // Revalidate authorization on every operation, not only first bind.
      const fresh = this.#helloExchange.reliableChannelContext(devicePath);
      try {
        if (fresh.peerTrustId !== this.#peerTrustId) {
          this.reset();
          fail(
            "PEER_TRUST_MISMATCH",
            "bound session changed peer trust context"
          );
        }
        return this.#channel;
      } finally {
        fresh.material.clientToServer.key.fill(0);
        fresh.material.clientToServer.noncePrefix.fill(0);
        fresh.material.serverToClient.key.fill(0);
        fresh.material.serverToClient.noncePrefix.fill(0);
      }
    }
    const context = this.#helloExchange.reliableChannelContext(devicePath);
    try {
      this.#channel = new ReliableChannelV1({
        transport: {
          send: async (frame) => {
            const decoded = decodeReliableFragmentV1(frame);
            const transmitter =
              decoded.type === RELIABLE_FRAME_TYPES.ACK
                ? GATT_RELIABLE_TRANSMITTERS.ACK
                : GATT_RELIABLE_TRANSMITTERS.DATA;
            if (!this.#subscriptions.has(transmitter)) {
              fail(
                "TRANSMITTER_NOT_SUBSCRIBED",
                `${transmitter} is not subscribed`
              );
            }
            if (this.#publish === null) {
              fail("PUBLISHER_NOT_READY", "data-plane publisher is not bound");
            }
            await this.#publish({ transmitter, value: Buffer.from(frame) });
            this.#publishedFragments += 1;
          }
        },
        store: this.#store,
        peerTrustId: context.peerTrustId,
        mtu: context.mtu,
        txKey: context.material.serverToClient.key,
        rxKey: context.material.clientToServer.key,
        txNoncePrefix: context.material.serverToClient.noncePrefix,
        rxNoncePrefix: context.material.clientToServer.noncePrefix,
        onMessage: this.#onMessage,
        now: this.#now
      });
      this.#devicePath = devicePath;
      this.#peerTrustId = context.peerTrustId;
      this.#sessionBinds += 1;
      return this.#channel;
    } finally {
      context.material.clientToServer.key.fill(0);
      context.material.clientToServer.noncePrefix.fill(0);
      context.material.serverToClient.key.fill(0);
      context.material.serverToClient.noncePrefix.fill(0);
    }
  }

  #assertEnabled(): void {
    if (!this.#enabled) {
      fail("DATA_PLANE_DISABLED", "reliable GATT data plane is disabled");
    }
  }
}

File: V5BT-sorgenti-e-roadmap-rimanente-20260818/APPLICATIVI/Palmare/android-app/app/src/main/java/com/sentrapa/webkiosk/bluetooth/GattReliableDataPlaneV1.kt
package com.sentrapa.webkiosk.bluetooth

enum class GattReliableEndpointRoleV1 {
    CLIENT,
    SERVER
}

enum class GattReliableTransmitterV1 {
    DATA,
    ACK
}

class GattReliableChannelContextV1(
    val peerTrustId: String,
    val mtu: Int,
    val role: GattReliableEndpointRoleV1,
    val material: ReliableChannelMaterialV1
) : AutoCloseable {
    init {
        if (
            mtu !in ReliableFrameCodecV1.MINIMUM_GATT_MTU..
            ReliableFrameCodecV1.MAXIMUM_GATT_MTU ||
            !Regex("^[0-9a-f]{64}$").matches(peerTrustId) ||
            material.isClosed
        ) {
            material.close()
            gattReliableDataPlaneFailureV1(
                "INVALID_RELIABLE_CONTEXT",
                "reliable GATT context is not canonical"
            )
        }
    }

    override fun close() {
        material.close()
    }

    override fun toString(): String =
        "GattReliableChannelContextV1(mtu=$mtu, role=$role, material=<redacted>)"
}

fun interface GattReliableSessionContextProviderV1 {
    fun export(sessionToken: Long): GattReliableChannelContextV1
}

fun interface GattReliablePublisherV1 {
    fun publish(transmitter: GattReliableTransmitterV1, frame: ByteArray)
}

data class GattReliableDataPlaneSnapshotV1(
    val enabled: Boolean,
    val bound: Boolean,
    val dataSubscribed: Boolean,
    val ackSubscribed: Boolean,
    val receivedFragments: Long,
    val publishedFragments: Long,
    val sessionBinds: Long,
    val resets: Long,
    val failures: Long,
    val channel: ReliableChannelMetricsV1?
)

class GattReliableDataPlaneExceptionV1(
    val code: String,
    message: String,
    cause: Throwable? = null
) : RuntimeException(message, cause)

internal fun gattReliableDataPlaneFailureV1(
    code: String,
    message: String,
    cause: Throwable? = null
): Nothing = throw GattReliableDataPlaneExceptionV1(code, message, cause)

class GattReliableDataPlaneV1(
    private val contextProvider: GattReliableSessionContextProviderV1,
    private val store: ReliableChannelStoreV1,
    private val onMessage: ReliableChannelMessageHandlerV1,
    private val enabled: Boolean = false,
    private val now: () -> Long = System::currentTimeMillis
) : AutoCloseable {
    private val subscriptions = mutableSetOf<GattReliableTransmitterV1>()
    private var publisher: GattReliablePublisherV1? = null
    private var channel: ReliableChannelV1? = null
    private var boundSessionToken: Long? = null
    private var boundPeerTrustId: String? = null
    private var receivedFragments = 0L
    private var publishedFragments = 0L
    private var sessionBinds = 0L
    private var resets = 0L
    private var failures = 0L
    private var closed = false

    @Synchronized
    fun setPublisher(value: GattReliablePublisherV1?) {
        assertOpen()
        publisher = value
    }

    @Synchronized
    fun setSubscription(transmitter: GattReliableTransmitterV1, enabled: Boolean) {
        assertOpen()
        assertEnabled()
        if (enabled) subscriptions += transmitter else subscriptions -= transmitter
    }

    @Synchronized
    fun receive(
        sessionToken: Long,
        frame: ByteArray
    ): ReliableChannelReceiveResultV1 {
        assertOpen()
        assertEnabled()
        assertSubscriptionsReady()
        val activeChannel = bind(sessionToken)
        receivedFragments += 1
        return try {
            activeChannel.receiveFragment(frame)
        } catch (error: Throwable) {
            failures += 1
            throw error
        }
    }

    @Synchronized
    fun send(
        sessionToken: Long,
        input: ReliableChannelSendInputV1
    ): ReliableChannelSendResultV1 {
        assertOpen()
        assertEnabled()
        assertSubscriptionsReady()
        if (input.type == ReliableFrameTypeV1.ACK) {
            gattReliableDataPlaneFailureV1(
                "ACK_RESERVED",
                "ACK is emitted only by ReliableChannelV1"
            )
        }
        return try {
            bind(sessionToken).send(input)
        } catch (error: Throwable) {
            failures += 1
            throw error
        }
    }

    @Synchronized
    fun sendBound(input: ReliableChannelSendInputV1): ReliableChannelSendResultV1 {
        val token = boundSessionToken
            ?: gattReliableDataPlaneFailureV1(
                "RELIABLE_CHANNEL_NOT_BOUND",
                "reliable data plane has no authenticated bound peer"
            )
        return send(token, input)
    }

    @Synchronized
    fun restore(sessionToken: Long): Int {
        assertOpen()
        assertEnabled()
        assertSubscriptionsReady()
        return try {
            bind(sessionToken).restoreDurableOutbox()
        } catch (error: Throwable) {
            failures += 1
            throw error
        }
    }

    @Synchronized
    fun restoreBound(): Int {
        val token = boundSessionToken
            ?: gattReliableDataPlaneFailureV1(
                "RELIABLE_CHANNEL_NOT_BOUND",
                "reliable data plane has no authenticated bound peer"
            )
        return restore(token)
    }

    @Synchronized
    fun tick(): ReliableChannelTickResultV1 {
        assertOpen()
        assertEnabled()
        return channel?.tick() ?: ReliableChannelTickResultV1(0, 0, 0)
    }

    @Synchronized
    fun reset() {
        if (closed) return
        channel?.close()
        channel = null
        boundSessionToken = null
        boundPeerTrustId = null
        subscriptions.clear()
        resets += 1
    }

    @Synchronized
    fun snapshot(): GattReliableDataPlaneSnapshotV1 =
        GattReliableDataPlaneSnapshotV1(
            enabled = enabled,
            bound = channel != null,
            dataSubscribed = subscriptions.contains(GattReliableTransmitterV1.DATA),
            ackSubscribed = subscriptions.contains(GattReliableTransmitterV1.ACK),
            receivedFragments = receivedFragments,
            publishedFragments = publishedFragments,
            sessionBinds = sessionBinds,
            resets = resets,
            failures = failures,
            channel = channel?.snapshot()
        )

    @Synchronized
    override fun close() {
        if (closed) return
        reset()
        publisher = null
        closed = true
    }

    private fun bind(sessionToken: Long): ReliableChannelV1 {
        if (sessionToken <= 0) {
            gattReliableDataPlaneFailureV1(
                "INVALID_SESSION_TOKEN",
                "session token must be positive"
            )
        }
        val existing = channel
        if (existing != null) {
            if (boundSessionToken != sessionToken) {
                gattReliableDataPlaneFailureV1(
                    "SESSION_ARBITRATION_CONFLICT",
                    "one data plane cannot bind two session contexts"
                )
            }
            try {
                contextProvider.export(sessionToken).use { fresh ->
                    if (fresh.peerTrustId != boundPeerTrustId) {
                        gattReliableDataPlaneFailureV1(
                            "PEER_TRUST_MISMATCH",
                            "bound session changed peer trust context"
                        )
                    }
                }
            } catch (error: Throwable) {
                reset()
                if (error is GattReliableDataPlaneExceptionV1) throw error
                gattReliableDataPlaneFailureV1(
                    "RELIABLE_CHANNEL_NOT_AUTHORIZED",
                    "bound session authorization could not be revalidated",
                    error
                )
            }
            return existing
        }

        val context = try {
            contextProvider.export(sessionToken)
        } catch (error: Throwable) {
            gattReliableDataPlaneFailureV1(
                "RELIABLE_CHANNEL_NOT_AUTHORIZED",
                "reliable data requires one active authenticated session",
                error
            )
        }
        try {
            val clientToServerKey = context.material.clientToServer.key
            val clientToServerNonce = context.material.clientToServer.noncePrefix
            val serverToClientKey = context.material.serverToClient.key
            val serverToClientNonce = context.material.serverToClient.noncePrefix
            try {
                val clientRole = context.role == GattReliableEndpointRoleV1.CLIENT
                channel = ReliableChannelV1(
                    transport = ReliableChannelTransportV1(::publish),
                    peerTrustId = context.peerTrustId,
                    mtu = context.mtu,
                    txKey = if (clientRole) clientToServerKey else serverToClientKey,
                    rxKey = if (clientRole) serverToClientKey else clientToServerKey,
                    txNoncePrefix = if (clientRole) {
                        clientToServerNonce
                    } else {
                        serverToClientNonce
                    },
                    rxNoncePrefix = if (clientRole) {
                        serverToClientNonce
                    } else {
                        clientToServerNonce
                    },
                    onMessage = onMessage,
                    store = store,
                    now = now
                )
            } finally {
                clientToServerKey.fill(0)
                clientToServerNonce.fill(0)
                serverToClientKey.fill(0)
                serverToClientNonce.fill(0)
            }
            boundSessionToken = sessionToken
            boundPeerTrustId = context.peerTrustId
            sessionBinds += 1
            return checkNotNull(channel)
        } finally {
            context.close()
        }
    }

    private fun publish(frame: ByteArray) {
        val decoded = ReliableFrameCodecV1.decodeFragment(frame)
        val target = try {
            if (decoded.type == ReliableFrameTypeV1.ACK) {
                GattReliableTransmitterV1.ACK
            } else {
                GattReliableTransmitterV1.DATA
            }
        } finally {
            decoded.payload.fill(0)
        }
        if (!subscriptions.contains(target)) {
            gattReliableDataPlaneFailureV1(
                "TRANSMITTER_NOT_SUBSCRIBED",
                "$target transmitter is not subscribed"
            )
        }
        val activePublisher = publisher
            ?: gattReliableDataPlaneFailureV1(
                "PUBLISHER_NOT_READY",
                "data-plane publisher is not bound"
            )
        val copy = frame.copyOf()
        try {
            activePublisher.publish(target, copy)
            publishedFragments += 1
        } finally {
            copy.fill(0)
        }
    }

    private fun assertSubscriptionsReady() {
        if (!subscriptions.contains(GattReliableTransmitterV1.DATA)) {
            gattReliableDataPlaneFailureV1(
                "DATA_SUBSCRIPTION_REQUIRED",
                "data transmitter subscription is required"
            )
        }
        if (!subscriptions.contains(GattReliableTransmitterV1.ACK)) {
            gattReliableDataPlaneFailureV1(
                "ACK_SUBSCRIPTION_REQUIRED",
                "ack transmitter subscription is required"
            )
        }
    }

    private fun assertEnabled() {
        if (!enabled) {
            gattReliableDataPlaneFailureV1(
                "DATA_PLANE_DISABLED",
                "reliable GATT data plane is disabled"
            )
        }
    }

    private fun assertOpen() {
        if (closed) {
            gattReliableDataPlaneFailureV1(
                "DATA_PLANE_CLOSED",
                "reliable GATT data plane is closed"
            )
        }
    }
}

class GattReliableApplicationChannelV1(
    private val dataPlane: GattReliableDataPlaneV1
) : BluetoothShadowChannelV1, RouteAdvertisementChannelV1 {
    override fun send(input: ReliableChannelSendInputV1): ReliableChannelSendResultV1 =
        dataPlane.sendBound(input)
}
